Lazily start active link refresher task

bulletmark opened the pull request:
This is a simple change that merely starts the active-links refresher background periodic task when needed, i.e. when there are one or more links to be checked. When there are no active links, the task is stopped. For an application that never uses active-links, no background refresher task is ever started.

This means a well crafted nicegui application (which only uses bindable properties/dataclass etc) will not wear the overhead of an unnecessary task running every 0.1 secs.

evnchn commented on nicegui/binding.py:
```diff
@@ -172,6 +189,8 @@ def bind_from(self_obj: Any, self_name: str, other_obj: Any, other_name: str,
     if (id(other_obj), other_name) not in bindable_properties:
+        if not active_links:
+            _ActiveLinkRefresher.start()
         active_links.append((other_obj, other_name, self_obj, self_name, backward))
```
Since you're adding to `active_links` so you know you need to start the active link refresher. 

Combined with how `_ActiveLinkRefresher.start()` starts the task if not started already, you should be able to straight-up call `_ActiveLinkRefresher.start()` with no if-condition. Same goes for above.

evnchn reviewed (changes requested):
Some code logic and API compatibility issues. 

But no need to rush and fix them. Let's affirm the need for this PR first.

bulletmark commented:
I read the template and considered that my commit words already addressed everything asked so deleted the headers etc sorry.

### Motivation

See 2nd paragraph of opening comment.

### Implementation

See 1st paragraph of opening comment.

### Progress

- [x] I chose a meaningful title that completes the sentence: "If applied, this PR will..."
- [x] The implementation is complete.
- [x] Pytests are not necessary.
- [x] Documentation change is not necessary as no user facing effect.

Regarding your question about whether this PR is needed. I feel using a 0.1 sec async task to check variables frequently has a rather strong "code smell" about it, particularly when I learned that a user can easily write an the application so that task is not even needed. I originally considered to just set my "binding_refresh_interval" to something like 99999 but, even apart from the nicegui code you quote, the Python `asyncio` scheduler is still baring the cost of managing a task + timer overhead. So I looked at the nicegui code and found it is very trivial to change/improve as done here. The overhead of creating/deleting the task only occurs when the count goes to/from 0 and 1 which is infrequent and I think is far better than wearing the cost of a permanent redundant 0.1 sec periodic task.

I don't have any metrics although I did run the original main and my branch against [a test I found](https://github.com/zauberzeug/nicegui/issues/2335) where somebody was exercising thousands of bind additions and deletions just to confirm I didn't break anything and that performance was the same - which it was.

I know I could call `_ActiveLinkRefresher.start()` every time but just thought the single `if` test removed the redundant python function call overhead. I agree with your last comment that I should only call `_ActiveLinkRefresher.stop()` in `remove()`.

bulletmark commented:
I force pushed to change `binding._refresh_loop()` to `binding.refresh_loop()` and remove the test + potential call to `_ActiveLinkRefresher.start()` in `remove()`.

I left the tests of `active_links` in the code because that simple test removes the overhead of the function call and it is more clear to a reader of the surrounding code that the function is only required when the list of links becomes/goes non-zero.

evnchn commented on nicegui/binding.py:
```diff
@@ -74,6 +74,21 @@ def _refresh_step() -> None:
+    def stop(cls) -> None:
+        if cls.task:
+            cls.task.cancel()
+            cls.task = None
```
@bulletmark I think that, since you call the `start` and `stop` accurately, then you can get rid of the checks `if not cls.task:` and `if cls.task` entirely.

evnchn commented:
The point about type checker is correct. Anyways I think we can wait for more reviewers, especially for whether the performance gain is worth it. Thanks for your work!

evnchn commented:
> raise on using active link when binding interval set as None

I don't think we should raise. That could be quite annoying to debug. I think we should warn instead. 

And we'll activate the refresh loop lazily in a one-way manner (no means of deactivating) if an active link is used. So the refreshable case would not lead to rapid start-stop. 

How does that sound?

bulletmark commented:
@falkoschindler my original plan for a PR was to add an option to set `binding_refresh_interval=-1` so the user could disable the refresh, much as you are proposing but then thinking about it I thought it better to just enable/disable the loop as required. However, I agree with the last proposal here by @evnchn. By default the loop is disabled but enabled and kept running forever once any active link binding is created. That makes fastidious users like myself happy, as we won't have the refresh loop running every 0.1 sec, and you happy because it avoids the possible thrashing problem you raise. I will redo this PR if you agree.

evnchn commented:
@bulletmark I think you can get started right away if you have time. I "won the argument" before on not raising errors unless absolutely necessary, so I think I'll win this time as well? (Check https://github.com/zauberzeug/nicegui/issues/4237#issuecomment-2816084272)

falkoschindler commented:
@evnchn @bulletmark You're right. Raising would be a bit harsh. 👍🏻

bulletmark commented:
I implemented as last discussed here then rebased off latest main and cleaned up to one commit.

If/after this change is committed to main then I will also submit a separate PR to add a test for this change. Didn't want to include it here because that makes it awkward over the transition.

falkoschindler commented:
> I implemented as last discussed here

Nice! 👍🏻

> then rebased off latest main and cleaned up to one commit.

Please try not to force-push to a pull request. This makes it harder to keep track of already reviewed files and can cause conflicts with local checkouts of this branch. When merging we will squash all changes into a single commit anyway, so the history looks clean again.

> If/after this change is committed to main then I will also submit a separate PR to add a test for this change. Didn't want to include it here because that makes it awkward over the transition.

What transition do you mean? What exactly do you find awkward? Usually tests are an essential part of the PR.

bulletmark commented on nicegui/binding.py:
```diff
@@ -51,10 +51,12 @@ def _set_attribute(obj: object | Mapping, name: str, value: Any) -> None:
 async def refresh_loop() -> None:
     """Refresh all bindings in an endless loop."""
+    binding_refresh_interval = core.app.config.binding_refresh_interval
+    assert binding_refresh_interval is not None, 'binding_refresh_interval is None'
```
I don't like dereferrencing `core.app..` twice. If you don't really want the user to set the value to 0 or negative then this is a good place to limit check that so perhaps:
```python
    if (binding_refresh_interval := core.app.config.binding_refresh_interval) is None or binding_refresh_interval <= 0:
        binding_refresh_interval = 0.1
````

If you don't want to check for `<= 0` then same test but just remove the 2nd `or` part?

evnchn commented:
Oh that's right! I think I'll let Falko decide himself on the final call. I just want to ensure the tests actually pass last midnight 🙇🏻‍♂️

evnchn commented:
Oh that's right! I think I'll let Falko decide himself on the final call. I just want to ensure the tests actually pass last midnight 🙇🏻‍♂️

evnchn commented:
> It shows how ... could cause a major bug

I feel "sniped" by the contextvar 😭 ... 

It makes the function, essentially, "dependent on where it was called". 

Anyways https://github.com/zauberzeug/nicegui/pull/5438/commits/8de03fa0a95a2ecf055cf0e5ce0a089212fe1a0d is my take at solving the problem without messing with the contextvar by keeping where it was called. 

I hope you like this solution. If not, either of you can force-push it away `git reset --hard HEAD~1` `git push --force`

evnchn commented:
`assert interval is not None` fails on Py3.9 but not other versions. No idea why. 

https://github.com/zauberzeug/nicegui/pull/5438/commits/c655c62a80fa5493aabdf61c171c0725cd285116 fixes things symtomatically. 

And `_ensure_refresh_loop_is_running` is no more (name is no longer accurate as well). 

The commit has an advantage: The binding loop will not crash despite repeated attempts to set `core.app.config.binding_refresh_interval = None`. Just a bit more robust.

evnchn commented:
> But tests are green on my machine.

Are you on Py3.9? later versions just kinda got over the differing event loop issue.

falkoschindler commented:
5ea4559b8ac785b393bebcdbd66ce4c5fb121607 seems to work.
Ready to merge if you ask me. 👍🏻
